### PR TITLE
LB: fix randomization in host LB scheduler initialization

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: LB: fix randomization in host LB scheduler initialization
Additional Description:
Following up on the randomized picking introduced in #31592, this PR uses that for host-picking LB.
Prior to this PR, the initialization would randomize a number between [0, #hosts). However, it should've randomize by the number of entries in the "pick list".
For example, if there are 2 hosts, host_A with weight 99 and host_B with weight 1, the randomization will land in either index 0 or 1, whereas it should have been between 0 and 99 (as there are 100 entries in the "pick list").
This may cause a skew in the traffic distribution, especially in a fleet of Envoy, and when Envoy receive periodic assignment updates.

The change ensures that the starting point of which host to choose from is randomized and picked from some random starting point.
(This is similar to #32075 but for host-picking in a locality).

Risk Level: Medium - a bug fix that may impact current traffic flows.
Testing: Added unit test, and updated old ones.
Docs Changes: None.
Release Notes: Added.
Platform Specific Features: N/A.
Runtime guard: Added envoy.reloadable_features.edf_lb_host_scheduler_init_fix

### CODE REVIEW

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Comment by adisuissa** [comment]
Thanks @wbpcode!
Also assigning @htuch who worked on this originally (this is a much smaller PR compared to the previous one).
/assign @htuch

File: source/common/upstream/load_balancer_impl.cc [hunk 1]
```diff
@@ -1084,28 +1084,52 @@ void EdfLoadBalancerBase::refresh(uint32_t priority) {
...
-    // We should probably change this to refresh at all times. See the comment in
-    // BaseDynamicClusterImpl::updateDynamicHostList about this.
-    for (const auto& host : hosts) {
-      // We use a fixed weight here. While the weight may change without
-      // notification, this will only be stale until this host is next picked,
-      // at which point it is reinserted into the EdfScheduler with its new
-      // weight in chooseHost().
-      scheduler.edf_->add(hostWeight(*host), host);
+
+    // If there are no hosts or a single one, there is no need for an EDF scheduler
```
Comment (htuch) [inline source/common/upstream/load_balancer_impl.cc:1088]: Nit: probably safe but not runtime guarded.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

File: source/common/upstream/load_balancer_impl.cc (hunk 1 above)
Comment (adisuissa) [reply source/common/upstream/load_balancer_impl.cc]: Better be safe than sorry. Done.

**Review by wbpcode — APPROVED** [review]

**Pull request merged**